--- a/qmacstyle/main.mm	2017-04-04 22:38:50.000000000 +0200
+++ b/qmacstyle/main.mm	2017-04-06 23:08:47.000000000 +0200
@@ -42,7 +42,7 @@
 
 QT_BEGIN_NAMESPACE
 
-class QMacStylePlugin : public QStylePlugin
+class QAltMacStylePlugin : public QStylePlugin
 {
     Q_OBJECT
     Q_PLUGIN_METADATA(IID "org.qt-project.Qt.QStyleFactoryInterface" FILE "macstyle.json")
@@ -50,11 +50,13 @@
     QStyle *create(const QString &key);
 };
 
-QStyle *QMacStylePlugin::create(const QString &key)
+QStyle *QAltMacStylePlugin::create(const QString &key)
 {
     QMacAutoReleasePool pool;
-    if (key.compare(QLatin1String("macintosh"), Qt::CaseInsensitive) == 0)
+    if (key.compare(QLatin1String("macintosh"), Qt::CaseInsensitive) == 0
+        || key.compare(QLatin1String("aqua"), Qt::CaseInsensitive) == 0) {
         return new QMacStyle();
+    }
 
     return 0;
 }
