diff --git a/qmacstyle_mac.mm b/qmacstyle_mac.mm
index 4264a03f65aef6b9dca295a6e08364fac91fc731..703e02a70544e196562329bcd278d5f66d5a0d17 100644
--- a/qmacstyle_mac.mm
+++ b/qmacstyle_mac.mm
@@ -144,16 +144,16 @@ static QWindow *qt_getWindow(const QWidget *widget)
     return widget ? widget->window()->windowHandle() : 0;
 }
 
-@interface QT_MANGLE_NAMESPACE(NotificationReceiver) : NSObject {
+@interface QT_MANGLE_NAMESPACE(AltNotificationReceiver) : NSObject {
 QMacStylePrivate *mPrivate;
 }
 - (id)initWithPrivate:(QMacStylePrivate *)priv;
 - (void)scrollBarStyleDidChange:(NSNotification *)notification;
 @end
 
-QT_NAMESPACE_ALIAS_OBJC_CLASS(NotificationReceiver);
+QT_NAMESPACE_ALIAS_OBJC_CLASS(AltNotificationReceiver);
 
-@implementation NotificationReceiver
+@implementation AltNotificationReceiver
 - (id)initWithPrivate:(QMacStylePrivate *)priv
 {
     self = [super init];
@@ -2168,8 +2168,8 @@ QMacStyle::QMacStyle()
     Q_D(QMacStyle);
     QMacAutoReleasePool pool;
 
-    d->receiver = [[NotificationReceiver alloc] initWithPrivate:d];
-    NotificationReceiver *receiver = static_cast<NotificationReceiver *>(d->receiver);
+    d->receiver = [[AltNotificationReceiver alloc] initWithPrivate:d];
+    AltNotificationReceiver *receiver = static_cast<AltNotificationReceiver *>(d->receiver);
 
     [[NSNotificationCenter defaultCenter] addObserver:receiver
                                                       selector:@selector(scrollBarStyleDidChange:)
@@ -2193,7 +2193,7 @@ QMacStyle::~QMacStyle()
     [d->horizontalScroller release];
     [d->verticalScroller release];
 
-    NotificationReceiver *receiver = static_cast<NotificationReceiver *>(d->receiver);
+    AltNotificationReceiver *receiver = static_cast<AltNotificationReceiver *>(d->receiver);
     [[NSNotificationCenter defaultCenter] removeObserver:receiver];
     [receiver release];
 
