diff --git a/qmacstyle_mac.mm b/qmacstyle_mac.mm
index b35317c..7a7316d 100644
--- a/qmacstyle_mac.mm
+++ b/qmacstyle_mac.mm
@@ -684,7 +684,7 @@ static QSize qt_aqua_get_known_size(QStyle::ContentsType ct, const QWidget *widg
             else if (sz == QAquaSizeMini)
                 ret = QSize(-1, qt_mac_aqua_get_metric(kThemeMetricMiniPushButtonHeight));
 
-            if (!psh->icon().isNull()){
+            if (!psh->icon().isNull() && psh->text().isEmpty()){
                 // If the button got an icon, and the icon is larger than the
                 // button, we can't decide on a default size
                 ret.setWidth(-1);
@@ -3863,6 +3863,25 @@ void QMacStyle::drawControl(ControlElement ce, const QStyleOption *opt, QPainter
     case CE_ToolBoxTabShape:
         QCommonStyle::drawControl(ce, opt, p, w);
         break;
+    case CE_PushButton:
+        // adapted from QCommonStyle::drawControl()
+        if (const QStyleOptionButton *btn = qstyleoption_cast<const QStyleOptionButton *>(opt)) {
+            drawControl(CE_PushButtonBevel, btn, p, w);
+            QStyleOptionButton subopt = *btn;
+            if (!btn->icon.isNull() && !btn->text.isEmpty()) {
+                // remove the icon from our local QStyleOptionButton copy
+                subopt.icon = QIcon();
+            }
+            subopt.rect = subElementRect(SE_PushButtonContents, btn, w);
+            drawControl(CE_PushButtonLabel, &subopt, p, w);
+            if (btn->state & State_HasFocus) {
+                QStyleOptionFocusRect fropt;
+                fropt.QStyleOption::operator=(*btn);
+                fropt.rect = subElementRect(SE_PushButtonFocusRect, btn, w);
+                drawPrimitive(PE_FrameFocusRect, &fropt, p, w);
+            }
+        }
+        break;
     case CE_PushButtonBevel:
         if (const QStyleOptionButton *btn = qstyleoption_cast<const QStyleOptionButton *>(opt)) {
             if (!(btn->state & (State_Raised | State_Sunken | State_On)))
@@ -4049,8 +4068,9 @@ void QMacStyle::drawControl(ControlElement ce, const QStyleOption *opt, QPainter
             // tab. So, cheat a little here. However, if it *is* only an icon
             // the windows style works great, so just use that implementation.
             bool hasMenu = btn.features & QStyleOptionButton::HasMenu;
-            bool hasIcon = !btn.icon.isNull();
             bool hasText = !btn.text.isEmpty();
+            // do we have icon AND should it show (->empty text label since "ButtonsHaveIcons" is false on Mac)?
+            bool hasIcon = !btn.icon.isNull() && !hasText;
 
             if (!hasMenu && usingYosemiteOrLater) {
                 if (tds == kThemeStatePressed
