diff --git a/private/qstylehelper.cpp b/private/qstylehelper.cpp
index 24d90fc..141b731 100644
--- a/private/qstylehelper.cpp
+++ b/private/qstylehelper.cpp
@@ -278,7 +278,6 @@ void drawDial(const QStyleOptionSlider *option, QPainter *painter)
     buttonColor.setHsv(buttonColor .hue(),
                        qMin(140, buttonColor .saturation()),
                        qMax(180, buttonColor.value()));
-    QColor shadowColor(0, 0, 0, 20);
 
     if (enabled) {
         // Drop shadow
