diff --git a/cocoa/qcocoaintegration.mm b/cocoa/qcocoaintegration.mm
index 20805af..2a3e78c 100644
--- a/cocoa/qcocoaintegration.mm
+++ b/cocoa/qcocoaintegration.mm
@@ -626,7 +626,7 @@ QCocoaServices *QCocoaIntegration::services() const
 QVariant QCocoaIntegration::styleHint(StyleHint hint) const
 {
     if (hint == QPlatformIntegration::FontSmoothingGamma)
-        return 2.0;
+        return mOptions.testFlag(UseFreeTypeFontEngine)? 0.975 : 2.0;
 
     return QPlatformIntegration::styleHint(hint);
 }
