diff --git a/cocoa/qcocoatheme.mm b/cocoa/qcocoatheme.mm
index 4d74c11..5d047d9 100644
--- a/cocoa/qcocoatheme.mm
+++ b/cocoa/qcocoatheme.mm
@@ -319,7 +319,7 @@ QVariant QCocoaTheme::themeHint(ThemeHint hint) const
 {
     switch (hint) {
     case QPlatformTheme::StyleNames:
-        return QStringList(QStringLiteral("macintosh"));
+        return QStringList({QStringLiteral("aqua"), QStringLiteral("macintosh")});
     case QPlatformTheme::DialogButtonBoxLayout:
         return QVariant(QPlatformDialogHelper::MacLayout);
     case KeyboardScheme:

