diff --git a/cocoa/qcocoatheme.mm b/cocoa/qcocoatheme.mm
index 13af8fb..46c8b67 100644
--- a/cocoa/qcocoatheme.mm
+++ b/cocoa/qcocoatheme.mm
@@ -44,7 +44,6 @@
 
 #include <QtCore/QVariant>
 
-#include "qcocoacolordialoghelper.h"
 #include "qcocoafontdialoghelper.h"
 #include "qcocoasystemsettings.h"
 #include "qcocoasystemtrayicon.h"
@@ -64,6 +63,9 @@
 
 #ifdef QT_WIDGETS_LIB
 #include <QtWidgets/qtwidgetsglobal.h>
+#if QT_CONFIG(colordialog)
+#include "qcocoacolordialoghelper.h"
+#endif
 #if QT_CONFIG(filedialog)
 #include "qcocoafiledialoghelper.h"
 #endif
diff --git a/cocoa/qnswindowdelegate.mm b/cocoa/qnswindowdelegate.mm
index ce74aa9..a18fe69 100644
--- a/cocoa/qnswindowdelegate.mm
+++ b/cocoa/qnswindowdelegate.mm
@@ -78,6 +78,21 @@
     return m_cocoaWindow->screen()->availableGeometry().toCGRect();
 }
 
+#if !QT_MACOS_PLATFORM_SDK_EQUAL_OR_ABOVE(__MAC_10_11)
+/*
+    AppKit on OS X 10.10 wrongly calls windowWillUseStandardFrame:defaultFrame
+    from -[NSWindow _frameForFullScreenMode] when going into fullscreen, resulting
+    in black bars on top and bottom of the window. By implementing the following
+    method, AppKit will choose that instead, and resolve the right fullscreen
+    geometry.
+*/
+- (NSSize)window:(NSWindow *)window willUseFullScreenContentSize:(NSSize)proposedSize
+{
+    Q_ASSERT(NSEqualSizes(m_cocoaWindow->screen()->geometry().size().toCGSize(), proposedSize));
+    return proposedSize;
+}
+#endif
+
 - (BOOL)window:(NSWindow *)window shouldPopUpDocumentPathMenu:(NSMenu *)menu
 {
     Q_UNUSED(window);
